snakemake_storage_plugin_xrootd/__init__.py

```diff
@@ -16,178 +21,170 @@
-            # This is useful to create complex types from the user input.
-            "parse_func": parse_query_params,
-            # If a parse_func is specified, you also have to specify an unparse_func
-            # that converts the parsed value back to a string.
-            "unparse_func": unparse_query_params,
+            "help": "Download remote files",
+            "env_var": False,
             "required": False,
-            "nargs": "+",
+            "type": bool,
         },
     )
 
 
-# Required:
-# Implementation of your storage provider
-# This class can be empty as the one below.
-# You can however use it to store global information or maintain e.g. a connection
-# pool.
 class StorageProvider(StorageProviderBase):
-    # For compatibility with future changes, you should not overwrite the __init__
-    # method. Instead, use __post_init__ to set additional attributes and initialize
-    # futher stuff.
 
     def __post_init__(self):
-        # This is optional and can be removed if not needed.
-        # Alternatively, you can e.g. prepare a connection to your storage backend here.
-        # and set additional attributes.
-
-        userpass = ""
-        if self.settings.username:
-            userpass += self.provider.settings.username
-        if self.settings.password:
-            userpass += f":{self.settings.password}"
-        if userpass:
-            userpass += "@"
-        port = f":{self.settings.port}" if self.settings.port else ""
-        self.netloc = f"{userpass}{self.settings.host}{port}"
-
-        self.filesystem_client = client.FileSystem(f"root://{self.netloc}")
-        self.file_client = client.File()
+        # List of error codes that there is no point in retrying
+        self.no_retry_codes = [
+            3000,
+            3001,
+            3002,
+            3006,
+            3008,
+            3009,
+            3010,
+            3011,
+            3013,
+            3017,
+            3021,
+            3025,
+            3030,
+            3031,
+            3032,
+        ]
+
+    def _check_status(self, status: XRootDStatus, error_preamble: str):
+        if not status.ok:
+            if status.errno in self.no_retry_codes:
+                raise XRootDFatalException(f"{error_preamble}: {status.message}")
+            else:
+                raise IOError(f"{error_preamble}: {status.message}")
 
     @classmethod
     def example_queries(cls) -> List[ExampleQuery]:
         """Return an example queries with description for this storage provider (at
         least one)."""
         return [
             ExampleQuery(
-                query="root://eos/user/s/someuser/somefile.txt",
-                description="A file on an XrootD instance. Note that we omit the host "
-                "name from the query as that is given via the settings of the storage "
-                "provider.",
+                query="root://eosuser.cern.ch//eos/user/s/someuser/somefile.txt",
```

Comment:
Unsure about this change of the query url. Previously we had the host and the username configured in the settings. Does something speak against that approach?

Reply:
Many thanks for all the comments! I've sort of answered this above a bit but this is mostly to do with my experience of how I interact with XRootD and snakemake (and I think it is typical of how members of the LHCb collaboration use it but I can't speak more broadly than that).

Possibly instead it could make more sense to leave this settings in as options so that if someone wants to use them they can but don't have to specify them (certainly in our case having to specify them would generate extra work but I guess could save someone else a lot of time). I can try and ask around a bit some other people who I know have some more experience with XRootD to see what they think and get back to you!

Reply:
Also to answer your first question sorry, yes I could not get it to work as it is implemented already -- in LHCb at least we have pinned all our snakemake versions to `Snakemake<v8` for a while so that we can use the old remote storage provider instead (actually I could not even get it to import successfully previously)

Reply:
Ok, let us follow your lead here. Keeping the settings seems reasonable, but it should definitely work as well out of the box the way you use it at LHCb.